README
Ganglia-Gmetric-PP version 1.03
===============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by Six Apart Ltd

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Ganglia::Gmetric::PP.

1.03  Mon Dec  7 16:26:02 PST 2009
        - Make aggregator output using gmetric(1) to workaround protocol
          incompatibility with ganglia 3.1
1.02  Wed Oct 14 14:29:45 PDT 2009
        - Add overall timeout to aggregator test to avoid cpan-testers reported hang
1.01  Thu Oct  8 16:26:06 PDT 2009
        - Skip aggregator test if its dependencies are missing
1.00  Wed Sep 16 15:27:57 PDT 2009
        - Add more options for manipulating output metrics in aggregator
0.99  Tue Sep 15 16:38:28 PDT 2009
        - Fix daemonization issue
0.98  Tue Sep 15 14:49:19 PDT 2009
        - Options changes
0.97  Mon Sep 14 14:02:33 PDT 2009
        - Use sendto
0.96  Fri Sep 11 16:00:45 PDT 2009
        - First CPAN release
0.95  Fri Sep 11 14:34:58 PDT 2009
        - Add ->receive method
0.94  Fri Sep 11 13:36:20 PDT 2009
        - Add tests
0.93  Thu Sep 10 18:14:36 PDT 2009
        - Add docs and comments to aggregator. Don't require AnyEvent
0.91  Thu Sep 10 17:54:40 PDT 2009
        - Support Danga::Socket as fallback from AnyEvent
0.90  Thu Sep 10 17:36:50 PDT 2009
        - first working version
0.01  Tue Sep  1 23:43:33 2009
	- original version; created by h2xs 1.23 with options
		-AXc -n Ganglia::Gmetric::PP



t/gen-test-data.pl
#!/usr/bin/perl

use strict;
use warnings;

my @types = qw/ string int8 uint8 int16 uint16 int32 uint32 float double /;

my %data;
for my $type (@types) {
    my $f = "type.strace";
    my $name = "${type}test";
    my $value = rand 100;
    $value = int $value if $type =~ /int/;
    system "strace -s 1024 -o $f gmetric -t $type -n $name -v $value -u things";
    open my $fh, '<', $f or die $!;
    while (<$fh>) {
        next unless /^write/;
        #write(3, "\0\0\0\0\0\0\0\6double\0\0\0\0\0\ndoubletest\0\0\0\0\0\003123\0\0\0\0\6things\0\0\0\0\0\3\0\0\0<\0\0\0\0", 64) = 64
        die "badness" unless /("[^"]+")/;
        push @{ $data{$type} }, [eval $1, [$type, $name, $value]];
    }
    unlink $f;
}
use Data::Dumper;
$Data::Dumper::Useqq = 1;
$Data::Dumper::Indent = 1;
print Dumper(\%data);


lib/Ganglia/Gmetric/PP/XDR.pm
=head1 NAME

Ganglia::Gmetric::PP::XDR - Encoder/decoder of Sun RPC XDR data

=cut

package Ganglia::Gmetric::PP::XDR;

use base 'Exporter';

our @EXPORT_OK = qw(
    int32_to_xdr
    uint32_to_xdr
    float_to_xdr
    double_to_xdr
    string_to_xdr
    int16_to_xdr
    enum_to_xdr
    uint16_to_xdr
    xdr_to_int32
    xdr_to_uint32
    xdr_to_float
    xdr_to_double
    xdr_to_string
    xdr_to_int16
    xdr_to_enum
    xdr_to_uint16
);
our %EXPORT_TAGS = (
    ':all' => \@EXPORT_OK,
    ':encode' => [qw(
        int32_to_xdr
        uint32_to_xdr
        float_to_xdr
        double_to_xdr
        string_to_xdr
        int16_to_xdr
        enum_to_xdr
        uint16_to_xdr
    )],
    ':decode' => [qw(
        xdr_to_int32
        xdr_to_uint32
        xdr_to_float
        xdr_to_double
        xdr_to_string
        xdr_to_int16
        xdr_to_enum
        xdr_to_uint16
    )],
);

=for comment

XDR basic types:

int     32 bits, big endian
uint    32 bits, big endian
float   IEEE, big endian
double  IEEE, big endian
string  4 bytes of length followed by char data, padded with nulls to multiple of 4 bytes

=cut

# in 5.9 and up, endian-ness forcers are available for more types
use constant {
    XDR_INT     => 'N',
    XDR_UINT    => ($] > 5.009 ? 'l>' : 'l'),
    XDR_FLOAT   => ($] > 5.009 ? 'f>' : 'f'),
    XDR_DOUBLE  => ($] > 5.009 ? 'd>' : 'd'),
    XDR_STRING  => 'N/a*x![4]',
};

sub int32_to_xdr  { pack XDR_INT,    int $_[0] }
sub uint32_to_xdr { pack XDR_UINT,   int $_[0] }
sub float_to_xdr  { pack XDR_FLOAT,  $_[0] }
sub double_to_xdr { pack XDR_DOUBLE, $_[0] }
sub string_to_xdr { pack XDR_STRING, $_[0] }

*int16_to_xdr  = &int32_to_xdr;
*enum_to_xdr   = &int32_to_xdr;
*uint16_to_xdr = &uint32_to_xdr;

sub xdr_to_int32  { unpack XDR_INT,    $_[0] }
sub xdr_to_uint32 { unpack XDR_UINT,   $_[0] }
sub xdr_to_float  { unpack XDR_FLOAT,  $_[0] }
sub xdr_to_double { unpack XDR_DOUBLE, $_[0] }
sub xdr_to_string { unpack XDR_STRING, $_[0] }

*xdr_to_int16  = &xdr_to_int32;
*xdr_to_enum   = &xdr_to_int32;
*xdr_to_uint16 = &xdr_to_uint32;

1;


lib/Ganglia/Gmetric/PP/v3_0.pm
package Ganglia::Gmetric::PP::v3_0;

use strict;
use warnings;

use base 'Ganglia::Gmetric::PP';

=for comment

A gmetric UDP packet in ganglia 3.0 consists of:
* the Ganglia packet type identifier, which is the constant metric_user_defined (value 0).
* a Ganglia_gmetric_message structure

// ganglia-3.0.6/lib/protocol.x
struct Ganglia_gmetric_message {
  string type<>;
  string name<>;
  string value<>;
  string units<>;
  unsigned int slope;
  unsigned int tmax;
  unsigned int dmax;
};

XDR strings are encoded as a big-endian 4 byte integer count of characters in
the string, followed by character data, null-padded to a multiple of 4
characters. This is encodable with the Perl pack string (N/a*x![4]).

=cut

use constant {
    MAGIC_ID       => 0,
    GMETRIC_FORMAT => 'N(N/a*x![4])4N3',
};

sub send {
    my $self = shift;
    my @msg = (MAGIC_ID, @_);
    $msg[4] = Ganglia::Gmetric::PP::DEFAULT_UNITS unless defined $msg[4];
    $msg[5] = Ganglia::Gmetric::PP::DEFAULT_SLOPE unless defined $msg[5];
    $msg[6] = Ganglia::Gmetric::PP::DEFAULT_TMAX  unless defined $msg[6];
    $msg[7] = Ganglia::Gmetric::PP::DEFAULT_DMAX  unless defined $msg[7];
    $self->SUPER::send(pack GMETRIC_FORMAT, @msg);
}

sub parse {
    my @res = unpack GMETRIC_FORMAT, $_[1];
    die "bad magic" unless shift(@res) == $_[0]->MAGIC_ID;
    return @res;
}

1;


lib/Ganglia/Gmetric/PP/v3_1.pm
package Ganglia::Gmetric::PP::v3_1;

use strict;
use warnings;

use base 'Ganglia::Gmetric::PP';

=for comment

A gmetric UDP packet in ganglia 3.0 consists of:
* the Ganglia packet type identifier, which is the constant metric_user_defined (value 0).
* a Ganglia_gmetric_message structure

// ganglia-3.1.2/lib/gm_protocol.x
struct Ganglia_extra_data {
  string name<>;
  string data<>;
};

struct Ganglia_metadata_message {
  string type<>;
  string name<>;
  string units<>;
  unsigned int slope;
  unsigned int tmax;
  unsigned int dmax;
  struct Ganglia_extra_data metadata<>;
};

XDR strings are encoded as a big-endian 4 byte integer count of characters in
the string, followed by character data, null-padded to a multiple of 4
characters. This is encodable with the Perl pack string (N/a*x![4]).

=cut

use constant GMETRIC_FORMAT => '';
use constant {
    MAGIC_ID       => 133, # gmetric_string
    GMETRIC_FORMAT => 'N(N/a*x![4])3N3(N/a*x![4])2',
};

sub send {
    my $self = shift;
    my @msg = (Ganglia::Gmetric::PP::MAGIC_ID, @_);
    $msg[4] = Ganglia::Gmetric::PP::DEFAULT_UNITS unless defined $msg[4];
    $msg[5] = Ganglia::Gmetric::PP::DEFAULT_SLOPE unless defined $msg[5];
    $msg[6] = Ganglia::Gmetric::PP::DEFAULT_TMAX  unless defined $msg[6];
    $msg[7] = Ganglia::Gmetric::PP::DEFAULT_DMAX  unless defined $msg[7];
    $msg[8] = 0;
    $self->SUPER::send(pack GMETRIC_FORMAT, @msg);
}

sub parse {
    my @res = unpack GMETRIC_FORMAT, $_[1];
    die "bad magic" unless shift(@res) == $_[0]->MAGIC_ID;
    pop @res;
    return @res;
}

1;


lib/Ganglia/Gmetric/PP.pm
=head1 NAME

Ganglia::Gmetric::PP - Pure Perl emitter/parser of Ganglia gmetric monitoring
packets

=head1 SYNOPSIS

    my $gmetric = Ganglia::Gmetric::PP->new(host => 'localhost', port => 8649);
    $gmetric->send($type, $name, $value, $units, $slope, $tmax, $dmax);

=head1 DESCRIPTION

This module constructs Ganglia packets in the manner of the gmetric program and
sends them via UDP to a gmond. Additionally it can receive and/or parse gmetric
packets which is useful for constructing proxies. Though written in pure Perl
with no non-core dependencies, it tries to be fast.

=cut

package Ganglia::Gmetric::PP;

our $VERSION = '1.03';

use strict;
use warnings;

use base 'Exporter', 'IO::Socket::INET';

our @EXPORT_OK = qw(
    GANGLIA_VALUE_STRING
    GANGLIA_VALUE_UNSIGNED_CHAR
    GANGLIA_VALUE_CHAR
    GANGLIA_VALUE_UNSIGNED_SHORT
    GANGLIA_VALUE_SHORT
    GANGLIA_VALUE_UNSIGNED_INT
    GANGLIA_VALUE_INT
    GANGLIA_VALUE_FLOAT
    GANGLIA_VALUE_DOUBLE
    GANGLIA_SLOPE_ZERO
    GANGLIA_SLOPE_POSITIVE
    GANGLIA_SLOPE_NEGATIVE
    GANGLIA_SLOPE_BOTH
    GANGLIA_SLOPE_UNSPECIFIED
    METRIC_INDEX_TYPE
    METRIC_INDEX_NAME
    METRIC_INDEX_VALUE
    METRIC_INDEX_UNITS
    METRIC_INDEX_SLOPE
    METRIC_INDEX_TMAX
    METRIC_INDEX_DMAX
);
our %EXPORT_TAGS = (
    'all' => \@EXPORT_OK,
);

=head1 FUNCTIONS

=over 4

=item * $gmetric = Ganglia::Gmetric::PP->new(host => $host, port => $port)

Constructs a new object which sends gmetric packets to the specified C<host>
and UDP C<port>. If omitted, they default to localhost and 8649, respectively.

=item * $gmond = Ganglia::Gmetric::PP->new(listen_host => $host, listen_port => $port)

Constructs a new object which receives gmetric packets (e.g. in a gmond replacement).
If the $gmetric will be used for receiving packets, C<listen_host> and
C<listen_port> may be specified as well.

=cut

sub new {
    my $proto = shift;
    my $class = ref($proto) || $proto;

    my %p = @_;

    my %opts;

    $opts{LocalHost} = $p{listen_host} if $p{listen_host};
    $opts{LocalPort} = $p{listen_port} if $p{listen_port};

    unless (%opts) {
        $opts{PeerHost} = $p{host} || 'localhost';
        $opts{PeerPort} = $p{port} || 8649;
    }

    my $self = IO::Socket::INET->new(
        Proto    => 'udp',
        %opts,
    );

    die "failed to create socket: $!" unless $self;

    return bless $self, $class;
}

=item * $gmetric->send($type, $name, $value, $units, $slope, $tmax, $dmax)

Sends a Ganglia message. The parameters are:

=over 4

=item * $type

The type of data being sent. Must be one of these importable constants:

=over 4

=item * GANGLIA_VALUE_STRING

=item * GANGLIA_VALUE_UNSIGNED_CHAR

=item * GANGLIA_VALUE_CHAR

=item * GANGLIA_VALUE_UNSIGNED_SHORT

=item * GANGLIA_VALUE_SHORT

=item * GANGLIA_VALUE_UNSIGNED_INT

=item * GANGLIA_VALUE_INT

=item * GANGLIA_VALUE_FLOAT

=item * GANGLIA_VALUE_DOUBLE

=back

=item * $name

The name of the metric.

=item * $value

The current value of the metric.

=item * $units

A string describing the units of measure for the metric.

=item * $slope

A description of how the metric tends to change over time. Must be one of these importable constants:

=over 4

=item * GANGLIA_SLOPE_ZERO

Data is fixed, mostly unchanging.

=item * GANGLIA_SLOPE_POSITIVE

Value is always increasing (counter).

=item * GANGLIA_SLOPE_NEGATIVE

Value is always decreasing.

=item * GANGLIA_SLOPE_BOTH

Value can be anything.

=back

=item * $tmax

The maximum time in seconds between gmetric calls.

=item * $dmax

The lifetime in seconds of this metric.

=back

=cut

# exported constants. see http://code.google.com/p/embeddedgmetric/wiki/GmetricProtocol
use constant {
    GANGLIA_VALUE_STRING            => 'string',
    GANGLIA_VALUE_UNSIGNED_CHAR     => 'uint8',
    GANGLIA_VALUE_CHAR              => 'int8',
    GANGLIA_VALUE_UNSIGNED_SHORT    => 'uint16',
    GANGLIA_VALUE_SHORT             => 'int16',
    GANGLIA_VALUE_UNSIGNED_INT      => 'uint32',
    GANGLIA_VALUE_INT               => 'int32',
    GANGLIA_VALUE_FLOAT             => 'float',
    GANGLIA_VALUE_DOUBLE            => 'double',

    GANGLIA_SLOPE_ZERO              => 0, # data is fixed, mostly unchanging
    GANGLIA_SLOPE_POSITIVE          => 1, # is always increasing (counter)
    GANGLIA_SLOPE_NEGATIVE          => 2, # is always decreasing
    GANGLIA_SLOPE_BOTH              => 3, # can be anything
    GANGLIA_SLOPE_UNSPECIFIED       => 4,

    METRIC_INDEX_TYPE               => 0,
    METRIC_INDEX_NAME               => 1,
    METRIC_INDEX_VALUE              => 2,
    METRIC_INDEX_UNITS              => 3,
    METRIC_INDEX_SLOPE              => 4,
    METRIC_INDEX_TMAX               => 5,
    METRIC_INDEX_DMAX               => 6,
};

# internal constants
use constant {
    MAGIC_ID                        => 0,
    GMETRIC_FORMAT                  => 'N(N/a*x![4])4N3',

    DEFAULT_UNITS                   => '',
    DEFAULT_SLOPE                   => 3,
    DEFAULT_TMAX                    => 60,
    DEFAULT_DMAX                    => 0,
};

sub send {
    my $self = shift;
    my @msg = (MAGIC_ID, @_);
    $msg[4] = DEFAULT_UNITS unless defined $msg[4];
    $msg[5] = DEFAULT_SLOPE unless defined $msg[5];
    $msg[6] = DEFAULT_TMAX  unless defined $msg[6];
    $msg[7] = DEFAULT_DMAX  unless defined $msg[7];
    $self->SUPER::send(pack GMETRIC_FORMAT, @msg);
}

=item * @metric = $gmetric->receive()

Waits for a single gmetric packet on the UDP listen port and returns the parsed
metric (see C<parse>).

=cut

sub receive {
    my $self = shift;
    return() unless $self->recv(my $buf, 1 << 14);
    return $self->parse($buf);
}

=item * @metric = Ganglia::Gmetric::PP->parse($packet_data)

Parses a gmetric packet, which is typically received by a UDP server.

The elements returned match the arguments to C<send>:

    ($type, $name, $value, $units, $slope, $tmax, $dmax) = @metric;

This function may die if the given data does not resemble a gmetric packet.

=cut

sub parse {
    my @res = unpack GMETRIC_FORMAT, $_[1];
    die "bad magic" unless shift(@res) == MAGIC_ID;
    return @res;
}

1;

=back

=head1 AUTHOR

Adam Thomason, E<lt>athomason@cpan.org<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007-2009 by Six Apart, E<lt>cpan@sixapart.comE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut
